### PR TITLE
refactor package_info

### PR DESCRIPTION
**Pull request by pedia**
Upgrade to 0.2.0
Add class PackageInfo
Provide singleton method getInstance to fetch all information once.

### CODE REVIEW

File: packages/package_info/lib/package_info.dart [hunk 1]
```diff
@@ -4,19 +4,59 @@
...
+  static PackageInfo _instance;
 
-/// Returns the `bundleIdentifier` on iOS or `getPackageName` on Android
-Future<String> get packageName async =>
-    await _kChannel.invokeMethod('getPackageName');
+  static Future<PackageInfo> getInstance() async {
+    if (_instance == null) {
+      final Map<String, String> fromSystem =
+          await _kChannel.invokeMethod('getAll');
+      print("fromSystem $fromSystem");
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: Remove `print` statement.

File: packages/package_info/lib/package_info.dart [hunk 2]
```diff
@@ -4,19 +4,59 @@
...
+      assert(fromSystem != null);
+      assert(fromSystem["version"] != null);
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: Suggest we move these three asserts to the `PackageInfo` constructor.

File: packages/package_info/lib/package_info.dart [hunk 3]
```diff
@@ -4,19 +4,59 @@
...
+      _instance = new PackageInfo._(
+        version: fromSystem["version"],
+        buildNumber: fromSystem["buildNumber"],
+        packageName: fromSystem["packageName"],
+      );
+    }
+    return _instance;
+  }
+
+  /// property of the `CFBundleShortVersionString` on iOS or `versionName` on Android
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: Property DartDoc should start with uppercase letter; we already know it is a property. What we do need is a phrase that tells the user what is represented by the property = the thing that the two platform-specific pieces of information have in common. I suggest something like
```dart
/// The package version. `CFBundleShortVersionString` on iOS, `versionName` on Android.
```
Similarly for the other two properties.

File: packages/package_info/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+## 0.2.0
+
+* **Breaking change**. Change package_info to class PackageInfo.
```
Comment (mravn-google) [inline packages/package_info/CHANGELOG.md]: Should mention that this provides all package information with a single async call.

File: packages/package_info/lib/package_info.dart [hunk 5]
```diff
@@ -4,19 +4,59 @@
...
+/// ```dart
+///     PackageInfo paakcageInfo = await PackageInfo.getInstance()
+///     print("Version is: ${packcageInfo.version}");
+/// ```
+///
+class PackageInfo {
+  PackageInfo._(
+      {@required this.version,
+      @required this.buildNumber,
+      @required this.packageName})
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: A trailing comma should hopefully make `dartfmt` produce something prettier here.

File: packages/package_info/lib/package_info.dart [hunk 6]
```diff
@@ -4,19 +4,59 @@
...
+/// Wrap application Bundle (on iOS) and Package (on Android) information,
+/// providing cached verion, build number and package name.
+///
+/// PackgeInfo implemented in singleton, after first called
+/// [PackageInfo.getInstance], cached value never changed.
+///
+/// ```dart
+///     PackageInfo paakcageInfo = await PackageInfo.getInstance()
+///     print("Version is: ${packcageInfo.version}");
+/// ```
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: Please correct spelling mistakes in `packageInfo` variable name (twice).

File: packages/package_info/lib/package_info.dart [hunk 7]
```diff
@@ -4,19 +4,59 @@
...
+import 'package:meta/meta.dart';
 import 'package:flutter/services.dart';
 
-/// Provides information about the current application package.
 const MethodChannel _kChannel =
     const MethodChannel('plugins.flutter.io/package_info');
 
-/// Returns the `CFBundleShortVersionString` on iOS or `versionName` on Android
-Future<String> get version async => await _kChannel.invokeMethod('getVersion');
+/// Wrap application Bundle (on iOS) and Package (on Android) information,
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: Suggestions:

- Let the class-level DartDoc start out by saying what concept the class represents. Application package information, in this case. Then what that amounts to in platform-specific terms on iOS and Android.
- I find the mention of caching in lines 14,16-17 confusing. AFAIK, we are talking about static information retrieved from the application bundle. If the information could change at runtime, we should reconsider the API.
- Don't mention each piece of information here. We'll most likely forget to update it, if we add another piece. The needed documentation is already present, in more detail, on each property below.
- Avoid discussing the implementation. The fact that `PackageInfo` is a singleton is either obvious from the `getInstance()` API, or unimportant/slightly confusing as the `getInstance()` method is async.

File: packages/package_info/README.md [hunk 8]
```diff
@@ -1,19 +1,21 @@
...
+import 'package:package_info/package_info.dart';
 
-var version = await package_info.version;
-var buildNumber = await package_info.buildNumber;
-var packageName = await package_info.packageName;
+PackageInfo packageInfo = await PackageInfo.getInstance();
+
+String version = packageInfo.version;
+String buildNumber = packageInfo.buildNumber;
+String packageName = packageInfo.packageName;
```
Comment (mravn-google) [inline packages/package_info/README.md]: Here and everywhere below: Seems to me the `packageName` is now logically the main piece of information (even if it was added most recently). So we should reorder every place these three occur together, including constructor, property declarations and so on below.

File: packages/package_info/lib/package_info.dart [hunk 9]
```diff
@@ -4,19 +4,59 @@
...
+      assert(fromSystem["buildNumber"] != null);
+      assert(fromSystem["packageName"] != null);
+
+      _instance = new PackageInfo._(
+        version: fromSystem["version"],
+        buildNumber: fromSystem["buildNumber"],
+        packageName: fromSystem["packageName"],
+      );
+    }
+    return _instance;
```
Comment (mravn-google) [inline packages/package_info/lib/package_info.dart]: [nit] This implementation artificially wraps the `PackageInfo` instance in a newly created `Future` on each invocation of `getInstance()` following the first. We should be able to reuse the same `Future` instance on each call, e.g. by using a `Completer` or a second, private async method:

```dart
class PackageInfo {
  Future<PackageInfo> _instance;

  Future<PackageInfo> getInstance() {
    if (_instance == null) {
      Completer<PackageInfo> completer = new Completer<PackageInfo>();
      _kChannel.invokeMethod('getAll').then((dynamic result) {
        Map<String, String> fromSystem = result;
        completer.complete(new PackageInfo._(...));
      });
      _instance = completer.future;
    }
    return _instance;
  }
}
```

File: packages/package_info/lib/package_info.dart (hunk 3 above)
Comment (pedia) [reply packages/package_info/lib/package_info.dart]: OK, I will finish these in this week.

File: packages/package_info/lib/package_info.dart (hunk 7 above)
Comment (pedia) [reply packages/package_info/lib/package_info.dart]: dartdoc(from dart 1.24.3) always failed with:
```
Generation failed: Unknown type GenericFunctionTypeElementImpl
package:dartdoc/src/model.dart 1528                                  new ModelElement.from
package:dartdoc/src/model.dart 2607                                  new TopLevelVariable
package:dartdoc/src/model.dart 1306                                  Library._getVariables.<fn>
dart:core                                                            Iterable.toList
package:dartdoc/src/model.dart 1307                                  Library._getVariables
package:dartdoc/src/model.dart 1099                                  Library.constants
package:dartdoc/src/model.dart 1344                                  Library.allModelElements
package:dartdoc/src/model.dart 2396                                  Package.allModelElements.<fn>
dart:core                                                            List.forEach
package:dartdoc/src/model.dart 2395                                  Package.allModelElements
package:dartdoc/dartdoc.dart 171                                     DartDoc.generateDocs
/Users/whesse/dart/sdk/third_party/pkg/dartdoc/bin/dartdoc.dart 156  main.<fn>
package:stack_trace                                                  Chain.capture
/Users/whesse/dart/sdk/third_party/pkg/dartdoc/bin/dartdoc.dart 155  main
===== asynchronous gap ===========================
dart:async                                                           _Completer.completeError
package:dartdoc/dartdoc.dart 191                                     DartDoc.generateDocs
===== asynchronous gap ===========================
dart:async                                                           new Future.microtask
package:dartdoc/dartdoc.dart 130                                     DartDoc.generateDocs
/Users/whesse/dart/sdk/third_party/pkg/dartdoc/bin/dartdoc.dart 156  main.<fn>
===== asynchronous gap ===========================
dart:async                                                           new Future.microtask
/Users/whesse/dart/sdk/third_party/pkg/dartdoc/bin/dartdoc.dart 155  main.<fn>
package:stack_trace                                                  Chain.capture
/Users/whesse/dart/sdk/third_party/pkg/dartdoc/bin/dartdoc.dart 155  main
```

Can I fix the doc in next pull request?

**Comment by mravn-google** [comment]
@pedia Turns out some of the package info fields can be null. I've committed some updates.

**Pull request merged**